decisions/DR004-Core-language-support.md

```diff
@@ -0,0 +1,460 @@
+
+## Relevant data
+
+### Required language support
+
+`C++`, `Python`
+
+### Projected language support
+
+`C`, `JavaScript`, `Java`, `Swift`, `Rust`, `go`
+
+### Bindings
+
+Research has shown that the common denominator when addressing the above
+listed languages is a `C` interface. A standard `C` API surface, with
+clear and manageable ownership semantics would facilitate support in the
+broadest range of future programming environments.
+
+Ideally, any bindings would also allow idiomatic programming within
+the target language, rather than forcing specific design patterns from
+any internals of OpenAssetIO - such as synchronous programming, RIAA,
+etc.
+
+### Performance
+
+A core requirement raised by numerous parties is that the "happy path"
+should be as performant as possible. Thus, it is required that any
+mechanism introduced to broaden compatibility doesn't compromise the
+(current) majority deployment scenarios.
+
+Of note, a `C++` host talking to a `C++` manager has been the most
+frequently requested high-performance pathway.
+
+### Development lifecycle
+
+The overhead of maintaining dynamically linked `C++` libraries within a
+plugin-based architecture is often fraught with constantly chasing ABI
+changes and/or conflicts between mutual dependencies.
+
+Foundry has prior art in using `C` function pointer suites as a compiler
+isolation technique that can help de-couple the plugin host and the
+specific of any given plugin.
+
+In the constantly evolving landscape defining which dependency versions
+are in use within which versions of common DCC tools, avoiding the need
+to matrix-compile and juggle plugin search paths per-host would be
+massively beneficial.
+
+## Options considered
+
+### Option 1
+
+`C++` and `Python` support only.
+
+**Estimated Cost**: Small
+
+#### Pros
+
+-   Simplifies development.
+
+#### Cons
+
+-   Leaves use within languages other than these a potentially
+    challenging exercise left to the reader.
+
+### Option 2
+
+`C` core library, bound to `C++` and `Python` as appropriate.
+
+**Estimated Cost**: Large
+
+#### Pros
+
+-   Makes future bindings or adoption in `C`-based languages far
+    simpler as all core API functionality is inherently available
+    through idiomatic `C`.
+
+#### Cons
+
+-   The data model of OpenAssetIO _requires_ `map` and `optional` types.
+    Implementing these in pure `C` either requires significant coding,
+    or would add additional third-party dependencies. Either of these
+    add significantly to code complexity, maintenance overhead and
+    project risk.
+
+### Option 3
+
+`C++` core library, with critical business logic exposed via `C`
+interfaces. Bound to `Python`.
+
+**Estimated Cost**: Medium
+
+#### Pros
+
+-   The `C` API is still a first-class citizen within the architecture,
+    aiding future adoption in other languages.
+-   The core implementation can leverage the `C++17` standard library,
+    dramatically reducing code complexity.
+
+#### Cons
```

Comment:
Is that a con of the choice here, or just a con of that deployment scenario? It feels like more the former? I think we need to treat that scenario as a requirement, rather than an option, so it should probably show up as a con for options 2 instead - i.e. the addition of boilerplate to the happy path.

Reply:
Added a note that Option 3 natively supports the high-perf path [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5408c60dc1ec565784c68318eff67bd22701d105), and added a note on the potential overhead in Opt 2 [here](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/189/commits/5e678a5e6925b1a07500d21264baf2954518f70d).